auth.py

```diff
@@ -96,9 +96,17 @@ def login ():
         if not check_password (body ['password'], user ['password']):
             return 'invalid username/password', 403
 
+        # If the number of bcrypt rounds has changed, create a new hash.
+        new_hash = None
+        if config ['bcrypt_rounds'] != int (re.match ('\$2b\$\d+', user ['password']) [0].replace ('$2b$', '')):
```

Comment:
How about extracting this out to a function:

```python
def extract_bcrypt_rounds(hash):
  ...
```

To calculate this value, and add a couple of unit tests on it. Specifically, one we should have:

```python
def test_extract_rounds():
  salt = bcrypt.gensalt(rounds=7)
  self.assertEquals(7, extract_bcrypt_rounds(salt))
```

Other than that, looks good

Reply:
I wondered about that too. But because we use the function only in one place, I thought it was overkill to extract it, especially since the bcrypt hash format is so unchanging (single digit rounts are prepended with a 0, but the regex I use allows one or more digits, so it covers both cases anyway). But perhaps, for the sake of having unit tests, it could make sense. I'll amend the PR.

Reply:
> we use the function only in one place

It's not just about deduping. It's also very useful to be able to give a piece of code a *name*, and abstract away over the mechanics of what's begin done.

Instead of `random_string_manipulation_here(x)` it's a lot easier to be able to read WHAT the random string manipulation is supposed to do: `extract_meaningful_value_from_string(x)`

Reply:
And also testing, of course